### PR TITLE
refactor/envoyconfig-controller

### PR DESCRIPTION
**Pull request by roivaz**
This PR is a complete rewrite of the EnvoyConfig controller logic. There are no behaviour changes, which can assessed by the fact that the integration tests in `controllers/marin3r/envoyconfig_controller_suite_test.go` have almost no changes. Improvements are:

* All reconciling code moved to `pkg/reconcilers/marin3r/envoyconfig`
* The number of api calls per reconcile has been greatly reduced
* Each reconcile loop has 3 very clear differenciated phases: initialization, revisions reconcile and status reconcile
* Added subpackages `pkg/reconcilers/marin3r/envoyconfig/revisions`, `pkg/reconcilers/marin3r/envoyconfig/filters` with methods to get and list and sort revisions.
* All the code has unit tests

### CODE REVIEW

**Comment by roivaz** [comment]
@slopezz @raelga this PR can now be reviewed. As this is a complete rewrite of the controller I'd suggest you read directly files `controllers/marin3r/envoyconfig_controller.go` and then `pkg/reconcilers/marin3r/envoyconfig/reconcile_revisions.go` because is kind of hard to follow the diff. These two files hold most of the logic and I tried to write them in a way that it's very clear to the reader what is going on at each step. Let me know if you think something is not clear.

File: pkg/reconcilers/marin3r/envoyconfig/filters/revision_filters.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+)
+
+// RevisionFilter is an interface that revision label selectors
+// implement
+type RevisionFilter interface {
+	ApplyToLabelSelector(client.MatchingLabels)
+}
+
+// VersionFilter is a struct used to filter revisions
+// by theis version
```
Comment (slopezz) [inline pkg/reconcilers/marin3r/envoyconfig/filters/revision_filters.go:21]:
```suggestion
// by their version
```

**Review by slopezz — DISMISSED** [review]
Awesome job, just an small typo!

your `go fu` is starting to scare :stuck_out_tongue_winking_eye:

**Review by raelga — DISMISSED** [review]

**Review by slopezz — APPROVED** [review]
